### PR TITLE
[LLVM] Add `LLVM_full`, `LLVM_full_assert` 18 recipes

### PR DESCRIPTION
**Pull request by mofeing**

### CODE REVIEW

**Comment by Zentrik** [comment]
https://github.com/Zentrik/Yggdrasil/commit/25f1fbad83b73913119d0440f258da2ddc7e9293 should fix the patches not applying.

**Comment by Zentrik** [comment]
Odd, I could have sworn the patches applied when I build LLVM_full@18 x86_64-linux-gnu-cxx11 locally.

**Comment by Zentrik** [comment]
The symlink for the assert builds points to the LLVM 17 patches.

**Comment by giordano** [comment]
If it's an upstream bug, first open a PR upstream.

**Comment by Zentrik** [comment]
Filed https://github.com/llvm/llvm-project/issues/88664. 

Also, I noticed that while we download the 10.14 sdk here, that's missing `/usr/include/c++/v1` (as noted in https://github.com/phracker/MacOSX-SDKs/issues/21) and so doesn't get updated. Is this an issue?

**Comment by mofeing** [comment]
> Filed [llvm/llvm-project#88664](https://github.com/llvm/llvm-project/issues/88664).

Can you open an issue in https://github.com/JuliaLang/llvm-project?

**Comment by Zentrik** [comment]
I don't see a way to create issues, do you want me to make a pr with the fix?

**Comment by Zentrik** [comment]
The windows build seems to be failing on building `tblgen-to-irdl` in a similar fashion to https://github.com/llvm/llvm-project/issues/58015. We disabled building `tblgen-lsp-server` to deal with that.

**Comment by mofeing** [comment]
> I don't see a way to create issues, do you want me to make a pr with the fix?

Yes, sorry. I meant a PR.

**Comment by Zentrik** [comment]
We probably still need a patch to deal with `tblgen-to-irdl` failing on windows. I was trying to make sure it worked when building statically before opening an upstream issue but my pc crashed and I haven't been able to test that yet.

**Comment by giordano** [comment]
I'm sure there are still plenty of things to fix, as it's always the case with LLVM :slightly_smiling_face: But let's get this going and see where it fails.

File: L/LLVM/common.jl [hunk 1]
```diff
@@ -18,6 +18,7 @@ const llvm_tags = Dict(
     v"15.0.7" => "2593167b92dd2d27849e8bc331db2072a9b4bd7f", # julia-15.0.7-10
     v"16.0.6" => "499f87882a4ba1837ec12a280478cf4cb0d2753d", # julia-16.0.6-2
     v"17.0.6" => "0424864b3ac348210aab171668eea1a669956a31", # julia-17.0.6-3
+    v"18.1.5" => "6f748bf1d110dbf9a4ab7ff6dac261c64865d514", # julia-18.1.5-0
```
Comment (giordano) [inline L/LLVM/common.jl:21]: Reminder to self: we still need to create the tag, we'll do it when we're reasonably sure we've got all the patches we need.

**Comment by giordano** [comment]
Alright, Windows builds are (as usual) the only failing ones. [Example](https://buildkite.com/julialang/yggdrasil/builds/10081#018f39bc-da89-4158-a605-2efaa254356a/6-2350):
```
[00:26:06] ninja: job failed: : && /opt/bin/x86_64-w64-mingw32-libgfortran5-cxx11/x86_64-w64-mingw32-clang++ --target=x86_64-w64-mingw32 --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/ -remap -D__USING_SJLJ_EXCEPTIONS__ -D__CRT__NO_INLINE -pthread -DMLIR_CAPI_ENABLE_WINDOWS_DLL_DECLSPEC -fvisibility-inlines-hidden -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Werror=mismatched-tags -O3 -DNDEBUG -Wl,--stack,16777216    -Wl,--gc-sections tools/mlir/tools/tblgen-to-irdl/CMakeFiles/tblgen-to-irdl.dir/tblgen-to-irdl.cpp.obj tools/mlir/tools/tblgen-to-irdl/CMakeFiles/tblgen-to-irdl.dir/OpDefinitionsGen.cpp.obj -o bin/tblgen-to-irdl.exe -Wl,--out-implib,lib/libtblgen-to-irdl.dll.a -Wl,--major-image-version,0,--minor-image-version,0  lib/libLLVMTableGen.a  lib/libLLVMTableGen.a  lib/libMLIRIR.a  lib/libMLIRIRDL.a  lib/libMLIRTblgenLib.a  lib/libMLIRSupport.a  lib/libMLIRDialect.a  lib/libMLIRInferTypeOpInterface.a  lib/libMLIRSideEffectInterfaces.a  lib/libMLIRIR.a  lib/libMLIRSupport.a  lib/libLLVM-18jl.dll.a  lib/libLLVMTableGen.a  lib/libLLVMSupport.a  -lpsapi  -lshell32  -lole32  -luuid  -ladvapi32  -lws2_32  /workspace/destdir/lib/libz.dll.a  lib/libLLVMDemangle.a  lib/libMLIRTableGen.a  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && :
[00:26:06] clang-16: warning: argument unused during compilation: '-remap' [-Wunused-command-line-argument]
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::parser<long long>::parse(llvm::cl::Option&, llvm::StringRef, llvm::StringRef, long long&)
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::basic_parser_impl::getOptionWidth(llvm::cl::Option const&) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::basic_parser_impl::printOptionInfo(llvm::cl::Option const&, unsigned long long) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::parser<long long>::printOptionDiff(llvm::cl::Option const&, long long, llvm::cl::OptionValue<long long>, unsigned long long) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::Option::anchor()
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::Option::addOccurrence(unsigned int, llvm::StringRef, llvm::StringRef, bool)
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_ostream::operator<<(double)
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::ParseCommandLineOptions(int, char const* const*, llvm::StringRef, llvm::raw_ostream*, char const*, bool)
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::Option::error(llvm::Twine const&, llvm::StringRef, llvm::raw_ostream&)
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::generic_parser_base::getOptionWidth(llvm::cl::Option const&) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::generic_parser_base::printGenericOptionDiff(llvm::cl::Option const&, llvm::cl::GenericOptionValue const&, llvm::cl::GenericOptionValue const&, unsigned long long) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::generic_parser_base::printOptionInfo(llvm::cl::Option const&, unsigned long long) const
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::cl::AddLiteralOption(llvm::cl::Option&, llvm::StringRef)
[00:26:06] >>> defined at libLLVMSupport.a(CommandLine.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::outs()
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::nulls()
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_ostream::operator<<(llvm::formatv_object_base const&)
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_fd_stream::raw_fd_stream(int, bool)
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_fd_ostream::~raw_fd_ostream()
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_ostream::changeColor(llvm::raw_ostream::Colors, bool, bool)
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: duplicate symbol: llvm::raw_ostream::resetColor()
[00:26:06] >>> defined at libLLVMSupport.a(raw_ostream.cpp.obj)
[00:26:06] >>> defined at libLLVM-18jl.dll
[00:26:06] 
[00:26:06] ld.lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
```
@Zentrik this is what you were referring to, right?

**Comment by Zentrik** [comment]
Yh, presumably we want to copy https://github.com/mofeing/Yggdrasil/blob/llvm-18/L/LLVM/LLVM_full@18/bundled/patches/0001-Disable-tblgen-lsp-server.patch for `tblgen-to-irdl`.

**Comment by Zentrik** [comment]
🎉

**Comment by gbaraldi** [comment]
I guess they are linking to LLVMSupport still even with dylibs on but yeah that fix is fine.

**Comment by gbaraldi** [comment]
LGTM

File: L/LLVM/common.jl (hunk 1 above)
Comment (giordano) [reply L/LLVM/common.jl:21]: Tag created: https://github.com/JuliaLang/llvm-project/releases/tag/julia-18.1.5-0

**Comment by vchuravy** [comment]
I think folks may want `tblgen-to-irdl`. But generally we have to build tblgen on the host once and then provide it again.

**Comment by vtjnash** [comment]
Can we drop the last commit and fix the merge conflict, then merge, so that we don't block upgrading llvm for existing uses, for a conflict with building a new tool that seems to have been misconfigured in cmake somewhere?

**Comment by giordano** [comment]
Done.  @vchuravy @gbaraldi do you want to have another look?

**Comment by vchuravy** [comment]
Patch list is: https://github.com/JuliaLang/llvm-project/compare/llvmorg-18.1.7...julia-18.1.7-0

**Comment by giordano** [comment]
We recently did some cleanup of patches no longer needed in the release-18 branch :slightly_smiling_face:

**Pull request merged**